### PR TITLE
Enhance address space handling to accept numeric CIDR masks

### LINKED ISSUE #4733: Update address_space_size property to allow cidr subnets
**Is your feature request related to a problem? Please describe.**
currently address_space_size allows small (/24) ,medium (22) and large (16)

**Describe the solution you'd like**
When specifying subnet size via address_space_size 
I would like to just specify the cidr instead of small/medium/large and also use smaller subnets than 24 e.g 25, 26

**Describe alternatives you've considered**
I have considered adding additional sizes e.g xsmall but it doesn't make sense to me as someone who understands subnet masks.

**Additional context**
I plan to try implementing this and allowing backwards compatibility with the existing small, medium and large and add the ability to specify a custom cidr number.

### PR DESCRIPTION
**Pull request by JC-wk**
# Resolves #4733 #3431

## What is being addressed

Currently IP ranges are allocated based on t-shirt sizing s/m/l this introduces CIDR ranges to allow more granular subnet allocation

## How is this addressed

- Accepts cidr values as strings from "16" to "29" for subnet creation 
- Updates documentation
- Updates CHANGELOG.md
- Updates API version
- Adds Tests

### CODE REVIEW

**Comment by JC-wk** [comment]
I've tested allocating a /27 and that worked in my environment

image

**Comment by marrobi** [comment]
- `address_space_size=None` can raise unexpectedly in `WorkspaceRepository.get_address_space_based_on_size()`.
- The new code does `workspace_properties.get("address_space_size", "small").strip().lower()`, which assumes the value is always a string.
- If the key is present but the value is `None`, this becomes `None.strip()` and raises `AttributeError` before any defaulting can happen.
- Suggested fix: normalize more defensively, e.g. `raw_size = workspace_properties.get("address_space_size")` then `address_space_size = str(raw_size or "small").strip().lower()`, or explicitly reject `None` with `InvalidInput`.
- This is worth fixing because the method is intended to default to `"small"` when no meaningful value is supplied, and this PR is already broadening accepted input handling.
- I also suggest adding a test for `address_space_size = None` to lock in the intended behavior.
- I did not find any other validated change/improvement comments; the implementation, docs, changelog, and added tests are otherwise consistent.

**Comment by JC-wk** [comment]
Addressed feedback from @marrobi

**Comment by marrobi** [comment]
Still getting:

- Unrecognized non-empty values still silently default to `/24`

File: api_app/db/repositories/workspaces.py [hunk 1]
```diff
@@ -160,18 +160,36 @@ def automatically_create_application_registration(self, workspace_properties: di
 
     async def get_address_space_based_on_size(self, workspace_properties: dict):
         # Default the address space to 'small' if not supplied.
-        address_space_size = workspace_properties.get("address_space_size", "small").lower()
+        raw_size = workspace_properties.get("address_space_size")
+        if raw_size is None or str(raw_size).strip() == "":
+            address_space_size = "small"
+        else:
+            address_space_size = str(raw_size).strip().lower()
```
[the remark this answers is not in the document]
Comment (JC-wk) [reply api_app/db/repositories/workspaces.py:168]: templates updated

File: templates/workspaces/base/template_schema.json [hunk 2]
```diff
@@ -49,9 +49,18 @@
     "address_space_size": {
       "type": "string",
       "title": "Address space size",
-      "description": "Network address size (small, medium, large or custom) to be used by the workspace.",
-      "default": "small",
+      "description": "Network address size cidr or (small /24, medium /22, large /16 or custom with an IP) to be used by the workspace.",
+      "default": "24",
       "enum": [
+        "24",
+        "23",
```
[the remark this answers is not in the document]
Comment (JC-wk) [reply templates/workspaces/base/template_schema.json]: "25"–"29" are not suitable for workspaces but can be used by workspace services

File: templates/workspaces/unrestricted/template_schema.json [hunk 3]
```diff
@@ -41,9 +41,18 @@
     "address_space_size": {
       "type": "string",
       "title": "Address space size",
-      "description": "Network address size (small, medium, large or custom) to be used by the workspace.",
-      "default": "small",
+      "description": "Network address size cidr or (small /24, medium /22, large /16 or custom with an IP) to be used by the workspace.",
+      "default": "24",
       "enum": [
+        "24",
+        "23",
```
[the remark this answers is not in the document]
Comment (JC-wk) [reply templates/workspaces/unrestricted/template_schema.json]: "25"–"29" are not suitable for workspaces but can be used by workspace services

File: templates/workspaces/airlock-import-review/template_schema.json [hunk 4]
```diff
@@ -33,9 +33,18 @@
     "address_space_size": {
       "type": "string",
       "title": "Address space size",
-      "description": "Network address size (small, medium, large or custom) to be used by the workspace.",
-      "default": "small",
+      "description": "Network address size cidr or (small /24, medium /22, large /16 or custom with an IP) to be used by the workspace.",
+      "default": "24",
       "enum": [
+        "24",
+        "23",
```
[the remark this answers is not in the document]
Comment (JC-wk) [reply templates/workspaces/airlock-import-review/template_schema.json]: "25"–"29" are not suitable for workspaces but can be used by workspace services

**Comment by JC-wk** [comment]
@marrobi I think copilot is mostly happy, maybe check opus again?

**Comment by marrobi** [comment]
@JC-wk still getting:

The backend accepts 16–29 and the docs advertise up to /29, but the schema enum tops out at /24. Values like "25"–"29" will be rejected by JSON-schema validation before ever reaching the backend — directly defeating the "smaller than /24" goal from issue #4733. The Copilot reviewer flagged this on all three schemas and it remains unresolved.

Also, the new schema description text says "custom with an IP", but custom requires a CIDR (e.g. 10.2.1.0/25), not a plain IP. Minor wording fix worth making for consistency with the more accurate description in workspace_template.py / the docs.

**Comment by JC-wk** [comment]
> @JC-wk still getting:
> 
> The backend accepts 16–29 and the docs advertise up to /29, but the schema enum tops out at /24. Values like "25"–"29" will be rejected by JSON-schema validation before ever reaching the backend — directly defeating the "smaller than /24" goal from issue #4733. The Copilot reviewer flagged this on all three schemas and it remains unresolved.
> 
> Also, the new schema description text says "custom with an IP", but custom requires a CIDR (e.g. 10.2.1.0/25), not a plain IP. Minor wording fix worth making for consistency with the more accurate description in workspace_template.py / the docs.

Thanks @marrobi 
1) is by design, workspaces need /24 as a minimum but workspace_service templates  can request an IP Subnet smaller than /24. Have now changed the description to be /24 instead of /25

2) wording was already like that prior to this change, but now updated to 
"Network address size cidr or (small /24, medium /22, large /16 or custom with an IP range e.g. 10.2.1.0/25) to be used by the workspace."

File: templates/workspaces/airlock-import-review/template_schema.json [hunk 5]
```diff
@@ -33,9 +33,18 @@
     "address_space_size": {
       "type": "string",
       "title": "Address space size",
-      "description": "Network address size (small, medium, large or custom) to be used by the workspace.",
+      "description": "Network address size cidr or (small /24, medium /22, large /16 or custom with an IP range e.g. 10.2.1.0/25) to be used by the workspace.",
       "default": "small",
```
[the remark this answers is not in the document]
Comment (JC-wk) [reply templates/workspaces/airlock-import-review/template_schema.json:37]: updated

File: templates/workspaces/unrestricted/template_schema.json [hunk 6]
```diff
@@ -41,9 +41,18 @@
     "address_space_size": {
       "type": "string",
       "title": "Address space size",
-      "description": "Network address size (small, medium, large or custom) to be used by the workspace.",
+      "description": "Network address size cidr or (small /24, medium /22, large /16 or custom with an IP range e.g. 10.2.1.0/25) to be used by the workspace.",
       "default": "small",
```
[the remark this answers is not in the document]
Comment (JC-wk) [reply templates/workspaces/unrestricted/template_schema.json:45]: updated

File: templates/workspaces/base/template_schema.json [hunk 7]
```diff
@@ -49,9 +49,18 @@
     "address_space_size": {
       "type": "string",
       "title": "Address space size",
-      "description": "Network address size (small, medium, large or custom) to be used by the workspace.",
+      "description": "Network address size cidr or (small /24, medium /22, large /16 or custom with an IP range e.g. 10.2.1.0/25) to be used by the workspace.",
       "default": "small",
```
[the remark this answers is not in the document]
Comment (JC-wk) [reply templates/workspaces/base/template_schema.json:53]: updated

**Review by marrobi — APPROVED** [review]
LGTM if it works!

**Review by maxmartin-cgi — APPROVED** [review]

File: api_app/tests_ma/test_api/test_routes/test_workspaces.py [hunk 8]
```diff
@@ -803,6 +803,24 @@ async def test_post_workspace_services_creates_workspace_service_with_address_sp
...
+        workspace = sample_workspace()
+        workspace.properties["address_spaces"] = ["192.168.0.1/24"]
+        get_workspace_mock.return_value = workspace
+        basic_workspace_service_template.properties["address_space"] = "10.1.0.0/24"
+        create_workspace_service_item_mock.return_value = [sample_workspace_service(), basic_workspace_service_template]
+        workspace_service_input["properties"]["address_space_size"] = "15"
+
+        response = await client.post(app.url_path_for(strings.API_CREATE_WORKSPACE_SERVICE, workspace_id=WORKSPACE_ID), json=workspace_service_input)
+
+        assert response.status_code == status.HTTP_422_UNPROCESSABLE_CONTENT
```
[the remark this answers is not in the document]
Comment (JC-wk) [reply api_app/tests_ma/test_api/test_routes/test_workspaces.py:821]: false positive see https://fastapi.tiangolo.com/reference/status/#fastapi.status.HTTP_422_UNPROCESSABLE_CONTENT

[the remark this answers is not in the document]
Comment (JC-wk) [reply api_app/tests_ma/test_api/test_routes/test_workspaces.py:821]: Starlette 0.48.0 introduced HTTP_422_UNPROCESSABLE_CONTENT as a new name for HTTP_422_UNPROCESSABLE_ENTITY and deprecated the old name due to RFC 9110; see Kludex/starlette#2939.

File: api_app/api/routes/workspaces.py [hunk 9]
```diff
@@ -262,7 +262,10 @@ async def create_workspace_service(response: Response, workspace_service_input:
         # check workspace has address_spaces property
         if not workspace.properties.get("address_spaces"):
             raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=strings.WORKSPACE_DOES_NOT_HAVE_ADDRESS_SPACES_PROPERTY)
-        workspace_service.properties["address_space"] = await workspace_repo.get_address_space_based_on_size(workspace_service_input.properties)
+        try:
+            workspace_service.properties["address_space"] = await workspace_repo.get_address_space_based_on_size(workspace_service_input.properties)
```
[the remark this answers is not in the document]
Comment (SvenAelterman) [reply api_app/api/routes/workspaces.py]: @ChrisChapman-gh Is Copoilot of the mark here? If so, just resolve the comment, please.

**Review by SvenAelterman — APPROVED** [review]
